File: src/transition_strategy.py
"""
Transition strategy selector.

Given normalised [0,1] feature arrays from the outgoing track (at exit bar)
and incoming track (at entry bar), selects the most appropriate transition
type and parameters.  The same logic lives in the frontend
(frontend/src/utils/transitionStrategy.ts) — keep both in sync.
"""
from __future__ import annotations

from typing import List, Optional


def compute_transition_config(
    *,
    out_energy: Optional[List[float]] = None,
    out_bass: Optional[List[float]] = None,
    out_vocal: Optional[List[float]] = None,
    out_beat_strength: Optional[List[float]] = None,
    out_exit_bar: int = 0,
    in_energy: Optional[List[float]] = None,
    in_bass: Optional[List[float]] = None,
    in_vocal: Optional[List[float]] = None,
    in_beat_strength: Optional[List[float]] = None,
    in_entry_bar: int = 0,
    out_tempo: float = 120.0,
    fade_secs: float = 7.0,
) -> dict:
    """Rule-based transition selector.  Returns a dict matching TransitionConfig."""

    out_e = _avg(out_energy, out_exit_bar)
    out_b = _avg(out_bass, out_exit_bar)
    out_v = _avg(out_vocal, out_exit_bar)
    out_bs = _avg(out_beat_strength, out_exit_bar)

    in_e = _avg(in_energy, in_entry_bar)
    in_b = _avg(in_bass, in_entry_bar)
    in_v = _avg(in_vocal, in_entry_bar)
    in_bs = _avg(in_beat_strength, in_entry_bar)

    in_e_ahead = _avg(in_energy, in_entry_bar + 8, window=8)
    beat_time = 60.0 / max(out_tempo, 60)

    # 1. EQ Bass Swap
    if out_b > 0.5 and in_b > 0.5 and out_bs > 0.4 and in_bs > 0.4:
        return dict(
            type="eq_swap", curve="equal_power",
            fade_secs=fade_secs, bass_swap_at=0.3,
        )

    # 3. Filter Sweep (vocals on outgoing)
    if out_v > 0.5:
        return dict(
            type="filter_sweep", curve="equal_power",
            fade_secs=fade_secs,
            filter_start_hz=20000.0, filter_end_hz=200.0,
        )

    # 4. Echo Out (rhythmic content)
    if out_bs > 0.6 and out_e > 0.5:
        return dict(
            type="echo_out", curve="equal_power",
            fade_secs=fade_secs,
            delay_time_sec=round(beat_time, 3), delay_feedback=0.5,
        )

    # 5. Harmonic Blend (both low energy)
    if out_e < 0.35 and in_e < 0.35 and out_v < 0.3 and in_v < 0.3:
        return dict(
            type="harmonic_blend", curve="equal_power",
            fade_secs=max(fade_secs, 10.0),
        )

    # 6. Default
    return dict(type="crossfade", curve="equal_power", fade_secs=fade_secs)


def _avg(arr: Optional[List[float]], center: int, window: int = 4) -> float:
    if not arr:
        return 0.0
    lo = max(0, center - window)
    hi = min(len(arr), center + window)
    vals = arr[lo:hi]
    return sum(vals) / len(vals) if vals else 0.0

File: frontend/src/utils/transitionStrategy.ts
/**
 * Transition strategy selector.
 *
 * Given feature arrays from the outgoing and incoming tracks at their
 * transition points, selects the most appropriate transition type.
 * Runs entirely client-side using normalised [0,1] feature arrays
 * already present in PredictResponse.
 */
import type { PredictResponse, TransitionConfig } from '../types'

export function computeTransitionConfig(
  outPred: PredictResponse,
  outExitSec: number,
  inPred: PredictResponse,
  inEntrySec: number,
  outTempo: number,
  fadeSecs: number,
): TransitionConfig {
  const outBar = findBarIndex(outPred.bar_times, outExitSec)
  const inBar = findBarIndex(inPred.bar_times, inEntrySec)

  const outEnergy = avg(outPred.energy, outBar)
  const outBass = avg(outPred.bass_energy, outBar)
  const outVocal = avg(outPred.vocal_presence, outBar)
  const outBeat = avg(outPred.beat_strength, outBar)

  const inEnergy = avg(inPred.energy, inBar)
  const inBass = avg(inPred.bass_energy, inBar)
  const inVocal = avg(inPred.vocal_presence, inBar)
  const inBeat = avg(inPred.beat_strength, inBar)

  const beatTime = 60 / Math.max(outTempo, 60)

  // 1. EQ Bass Swap: both tracks have strong kicks
  if (outBass > 0.5 && inBass > 0.5 && outBeat > 0.4 && inBeat > 0.4) {
    return {
      type: 'eq_swap',
      curve: 'equal_power',
      fadeSecs,
      bassSwapAt: 0.3,
    }
  }

  // 3. Filter Sweep: outgoing has vocals
  if (outVocal > 0.5) {
    return {
      type: 'filter_sweep',
      curve: 'equal_power',
      fadeSecs,
      filterStartHz: 20000,
      filterEndHz: 200,
    }
  }

  // 4. Echo Out: strong rhythmic content at exit
  if (outBeat > 0.6 && outEnergy > 0.5) {
    return {
      type: 'echo_out',
      curve: 'equal_power',
      fadeSecs,
      delayTimeSec: round3(beatTime),
      delayFeedback: 0.5,
    }
  }

  // 5. Harmonic Blend: both tracks low energy, no vocals
  if (outEnergy < 0.35 && inEnergy < 0.35 && outVocal < 0.3 && inVocal < 0.3) {
    return {
      type: 'harmonic_blend',
      curve: 'equal_power',
      fadeSecs: Math.max(fadeSecs, 10),
    }
  }

  // 6. Default: equal-power crossfade
  return { type: 'crossfade', curve: 'equal_power', fadeSecs }
}

// ── helpers ──────────────────────────────────────────────────────────────

function findBarIndex(barTimes: number[] | undefined, sec: number): number {
  if (!barTimes?.length) return 0
  let best = 0
  let bestDist = Math.abs(barTimes[0] - sec)
  for (let i = 1; i < barTimes.length; i++) {
    const d = Math.abs(barTimes[i] - sec)
    if (d < bestDist) { bestDist = d; best = i }
  }
  return best
}

function avg(arr: number[] | undefined, center: number, window = 4): number {
  if (!arr?.length) return 0
  let sum = 0, count = 0
  const lo = Math.max(0, center - window)
  const hi = Math.min(arr.length, center + window)
  for (let i = lo; i < hi; i++) { sum += arr[i]; count++ }
  return count > 0 ? sum / count : 0
}

function round3(n: number): number { return Math.round(n * 1000) / 1000 }
